content/en/blog/_posts/2018-09-19-network-bootable-farm-with-ltsp.md

```diff
@@ -0,0 +1,485 @@
+      grub-common \
+      grub-pc-bin \
+      grub-efi-amd64-bin \
+      curl \
+      patch \
+ && sed -i 's|in_target mount|in_target_nofail mount|' \
+      /usr/share/debootstrap/functions \
+  # Add EFI support and Grub bootloader (#1745251)
+ && patch -p2 -d /usr/sbin < /patches/feature-grub.diff \
+ && rm -rf /var/lib/apt/lists \
+ && apt-get clean
+```
+
+At this stage our docker image already have installed:
+
+* nbd-server
+* tftp-server
+* ltsp-scripts with grub bootloader support (for EFI)
+
+### Stage 2: basesystem
+
+In this stage we will prepare chroot environment with basesystem, and install basic software with the kernel there.
+We will use classic **debootstrap** instead **ltsp-build-client** for prepare base image, because **ltsp-build-client** will install GUI and few other things which we don't need for the server deployment.
+
+```Dockerfile
+FROM ltsp-base as basesystem
+
+ARG DEBIAN_FRONTEND=noninteractive
+
+# Prepare base system
+RUN debootstrap --arch amd64 xenial /opt/ltsp/amd64
+
+# Install updates
+RUN echo "\
+      deb http://archive.ubuntu.com/ubuntu xenial main restricted universe multiverse\n\
+      deb http://archive.ubuntu.com/ubuntu xenial-updates main restricted universe multiverse\n\
+      deb http://archive.ubuntu.com/ubuntu xenial-security main restricted universe multiverse" \
+      > /opt/ltsp/amd64/etc/apt/sources.list \
+ && ltsp-chroot apt-get -y update \
+ && ltsp-chroot apt-get -y upgrade
+
+# Installing LTSP-packages
+RUN ltsp-chroot apt-get -y install ltsp-client-core
+
+# Apply initramfs patches
+# 1: Read params from /etc/lts.conf during the boot (#1680490)
+# 2: Add support for PREINIT variables in lts.conf
+ADD /patches /patches
+RUN patch -p4 -d /opt/ltsp/amd64/usr/share < /patches/feature_initramfs_params_from_lts_conf.diff \
+ && patch -p3 -d /opt/ltsp/amd64/usr/share < /patches/feature_preinit.diff
+
+# Write new local client config for boot nbd image to ram:
+RUN echo "[Default]\nLTSP_NBD_TO_RAM = true" \
+      > /opt/ltsp/amd64/etc/lts.conf
+
+# Install packages
+RUN echo 'APT::Install-Recommends "0";\nAPT::Install-Suggests "0";' \
+      >> /opt/ltsp/amd64/etc/apt/apt.conf.d/01norecommend \
+ && ltsp-chroot apt-get -y install \
+      software-properties-common \
+      apt-transport-https \
+      ca-certificates \
+      ssh \
+      bridge-utils \
+      pv \
+      jq \
+      vlan \
+      bash-completion \
+      screen \
+      vim \
+      mc \
+      lm-sensors \
+      htop \
+      jnettop \
+      rsync \
+      curl \
+      wget \
+      tcpdump \
+      arping \
+      apparmor-utils \
+      nfs-common \
+      telnet \
+      sysstat \
+      ipvsadm \
+      ipset \
+      make
+
+# Install kernel
+RUN ltsp-chroot apt-get -y install linux-generic-hwe-16.04
+```
+
+Be careful since kernel installed some packages like `lvm2` may have problems with installing in unprivileged chroot, because their postinstall scripts try to handle some privileged commands.
```

Comment:
> Be careful in cases where the Linux kernel has some packages like `lvm2` installed. This can cause problems when installing in unprivileged chroot, because post-install scripts may attempt to handle some privileged commands.

Reply:
Here I meant that if you will install lvm2 before kernel you will have not this issue, because it won't be call those privileged commands.
However, after the kernel installed such problems can arise

Reply:
Changed to:

>Note that you may encounter problems with some packages, such as `lvm2`.
>They have not fully optimized for installing in unprivileged chroot.
>Their postinstall scripts try to call some privileged commands which can fail with errors and block whole package installation.
>
>Solution:
>* Some of them can be simple installed before the kernel without any problems (like `lvm2`)
>* But for some of them you will need to use this workaround for install without postinstall script.